### PR TITLE
[feat] LoRA support on V1 (with limitations)

### LINKED ISSUE #166: [feature] LoRA support on V1
LoRA (Low-Rank Adaptation) support for vLLM-RBLN V1 on RBLN hardware.

### Graph Compilation Constraints

**Pre-compilation requirement**: LoRA weights must be prepared before graph compilation
- Currently, there is no direct method to access device memory within the compiled graph at runtime
- All LoRA weight structures must be known during the compilation phase
- Dynamic LoRA loading after compilation is not supported now

### LoRA Masking Strategy

**Fixed Graph with Masking**: Always compute for N LoRA (for multi-lora) adapters with masking to apply only active ones
- Maintains consistent computation graph and tensor shapes across requests
- Uses masking mechanism to enable/disable specific LoRA adapters
- Accepts computational overhead to preserve graph stability
- Ensures predictable memory layout and execution flow

### Implementation Details

**Pure Matmul Implementation**: Initially implementing with standard matrix multiplication operations only
- No custom kernels in the first iteration
- **Future Consideration**: Custom kernel optimization may be needed for better performance

## Expected Limitations

- V1 only
- Single LoRA: Multi-LoRA not supported due to prefill scheduling constraints
- No warmup: Warmup must be disabled; first request must include LoRA request
- Static Compilation: Cannot add new LoRA adapters after graph compilation

### PR DESCRIPTION
**Pull request by junstar92**
### 🚀 Summary of Changes

Implements LoRA (Low-Rank Adaptation) support by creating a custom Punica Wrapper and platform-specific adaptations for RBLN hardware (**initial version**). It covers only a Single-LoRA.

#### Changes

Changes based on vLLM
- Implemented custom `PunicaWrapperRBLN` class
  - `add_lora_embedding`: LoRA for `VocabParallelEmbedding` layers
  - `add_lora_linear`: LoRA for linear layers (QKV, MLP, etc.)
  - `add_lora_logits`: LoRA for logits processor
- Pure matrix multiplication implementation without custom kernels

Platform (RBLN)-specific Changes
- LoRA Mask and Inputs
  - Added `LoRAMask` to make masking graph input
  - Added `LoRAInputs` for managing `sampler_indices_padded`
  - Required for adding LoRA-specific inputs to RBLN compilation graphs
- LoRA Data Type Setting: Cannot set `float32` by engine arguments due to vLLM's pydantic limitations
- Monkey Patching
  - `BaseLinearLayerWithLoRA.apply`: Modified to handle vLLM-RBLN tensor shapes
  - `VocabParallelEmbeddingWithLoRA.forward`: Adapted for RBLN embedding layer structure

Others
- Added an environment condition to skip warmup

#### Limitations

- **V1 Only**
- **Pre-compilation LoRA Weights**: LoRA weights must be prepared before graph compilation
  - Graph compilation requires knowing LoRA structure ahead of time
  - Cannot dynamically add new LoRA adapters after compilation
- **No Warmup Support**: Current warmup doesn't include LoRA information
  - Temporal Workaround: Disable warmup and add LoRA request in the first request
- **Single LoRA Only**: Multi-LoRA currently not supported
  - Current limitation: Only one prefill allowed per step
  - Cannot load multiple LoRA weights simultaneously for graph compilation
  - Multi-LoRA support requires architectural changes to prefill scheduling and kernel assertion
- **Fallback to eager execution** due to RBLM compilation failure
  - unsupported operations: `aten::nan_to_num_` and `aten::mT`
---

### 📌 Related Issues / Tickets

* Resolves #166 

---

### ✅ Type of Change

* [x] ✨ Feature (`feature`)
* [ ] 🧠 Model support (`model`)
* [ ] 🧬 Core engine changes (`core`)
* [ ] 🛠 Bug fix (`bug-fix`)
* [ ] ⚙️ Performance improvement (`perf`)
* [ ] 🔁 Refactor or code cleanup (`refactor`)
* [ ] 📄 Documentation (`docs`)
* [ ] ❓ Other (`other`): please describe

---

### 🧪 How to Test

You can test it simply with examples/experimental/run_lora_test.py
```
$ VLLM_RBLN_ENFORCE_MODEL_FP32=1 python examples/experimental/run_lora_test.py
```
Unit tests will be included in a separate PR.

---

### 📸 Screenshots / Logs (if applicable)

---

### 📋 Checklist

* [x] PR title follows Conventional Commits format
* [x] This PR is linked to an existing issue
* [ ] The test method is described, and the expected result is clearly stated
* [ ] Relevant documentation has been updated (if applicable)

---

### 💬 Notes

1. Is there special reason why the "remove padding in vocab" logic is placed in the `_gather_logits` method instead of the `_get_logits` method, unlike vLLM ? When LoRA is applied, `_get_logits` calls `_gather_logits`, but LoRA extends the vocab size. This logic doesn't account for the LoRA-extended vocab. Currently, I commented out these lines temporarily.
https://github.com/rebellions-sw/vllm-rbln/blob/ea0746e3c2a87e01a7f97756464032bfae13d575/vllm_rbln/model_executor/layers/logits_processor.py#L46-L48

2. I'm considering adding a simple LoRA example script. Any suggestions on where it should be placed? (e.g., `examples/experimental`, etc.)

---

### CODE REVIEW

**Comment by tomtomjhj** [comment]
> Is there special reason why the "remove padding in vocab" logic is placed in the _gather_logits method instead of the _get_logits method, unlike vLLM ?

cc @rebel-wonsubkim 

> I'm considering adding a simple LoRA example script. Any suggestions on where it should be placed? (e.g., examples/experimental, etc.)

Please put it in `examples/experimental/`

File: vllm_rbln/lora/inputs.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import torch
+
+
+class LoRAInputs:
+    sampler_indices_padded: torch.Tensor
+
+    @staticmethod
+    def set_sampler_indices_padded(
+            sampler_indices_padded: torch.Tensor) -> None:
+        LoRAInputs.sampler_indices_padded = sampler_indices_padded
```
Comment (rebel-jaehwang) [inline vllm_rbln/lora/inputs.py]: nit:
```suggestion
    sampler_indices_padded: ClassVar[torch.Tensor]

    @classmethod
    def set_sampler_indices_padded(
            cls, sampler_indices_padded: torch.Tensor) -> None:
        LoRAInputs.sampler_indices_padded = sampler_indices_padded
```
and similar changes for the get method and LoraMask class

**Review by rebel-jaehwang — APPROVED** [review]
lgtm with a minor nit

File: vllm_rbln/lora/inputs.py (hunk 1 above)
Comment (junstar92) [reply vllm_rbln/lora/inputs.py]: Just to confirm, is this what you intended?
